lunaform/names.go

```diff
@@ -0,0 +1,86 @@
+// Copyright 2017 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package lunaform
+
+import (
+	"crypto/rand"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+// regex for cleaning up host names, meets GCE's requirements.
+var nameFilterRe *regexp.Regexp
+
+func init() {
+	nameFilterRe = regexp.MustCompile(`[^-a-z0-9]`)
+}
+
+// strip any invalid characters and truncate
+func nameFilter(name string, limit int) string {
+	name = strings.ToLower(name)
+	name = nameFilterRe.ReplaceAllString(name, "")
+	if len(name) > limit {
+		name = name[:limit]
+	}
+	return name
+}
+
+// detect the process name
+func getProc() string {
+	var name string
+	if len(os.Args) >= 1 {
+		name = nameFilter(filepath.Base(os.Args[0]), 9)
+	}
+	// be extra fussy and require a letter at the beginning
+	if name == "" || name[0] < 'a' || name[0] > 'z' {
+		name = "lunaform"
+	}
+	return name
+}
+
+// detect user name
+func getUser() string {
+	name := nameFilter(os.Getenv("USER"), 9)
+	if name == "" {
+		name = "somebody"
+	}
+	return name
+}
+
+// generate something random
+func getRand() string {
+	b := make([]byte, 8)
+	// it is a pretty bad day if this fails, just bail.
+	if _, err := io.ReadFull(rand.Reader, b); err != nil {
+		panic(fmt.Errorf("random source failed: %v", err))
+	}
+	return fmt.Sprintf("%x", b)
+}
+
+// NamePrefix generates a hopefully unique resource name prefix that
+// includes the current process, user, and test names for context.
+// The result is at most 57 characters so a small suffix can be added
+// and still fit into the 63 character host name limit.
+func NamePrefix(testName string) string {
+	procName := getProc() // 9 chars max
+	userName := getUser() // 9 chars max
+	testName = nameFilter(testName, 20)
```

Comment:
Just a thought then, when using prefix matching for names, they end up kinda being structured like a directory tree for me with lots of names and dots instead of slashes so they get a bit long. I could be convinced to use something else, but do we need the test name in this? I mostly use the `basename` feature to help reap lost machines so I really mostly care about what and who launched the test rather then the actual test the machine was assigned to. For example, in pluton's case, we have different PR jobs launching different tests and so I like to distinguish which PR job launched the machines I see floating around to know if they are in-use. If a random developer launches the machines I will see they are the default name (so the automatic username is definitely an improvement here).

Of course, kola could also be smarter about intercepting signals (the most common way we have machines floating loose of a test) which, I imagine, would reduce this use-case greatly.